design/volume-snapshot-data-movement/volume-snapshot-data-movement.md

```diff
@@ -0,0 +1,779 @@
+
+**Set Backup As WaitForAsyncOperations**  
+**Persist Backup**  
+After ```Execute``` returns, the backup is set to ```WaitingForPluginOperations```, and then Velero is free to process other items or backups.  
+Before Velero moves to other items/backups, it will persist the backup data. This is the same as the existing behavior.  
+The backup then is left as ```WaitForAsyncOperations``` until the DM completes or timeout.  
+
+**BIA Progress**  
+Velero keeps monitoring the status of the backup by calling BIA V2’s Progress method. Below is the Progress method from BIA V2:
+```
+Progress(operationID string, backup *api.Backup) (velero.OperationProgress, error)
+```
+On the call of this method, DMP will query the DUCR’s status. Some critical progress data is transferred from DUCR to the ```OperationProgress``` which is the return value of BIA V2’s Progress method. For example, NCompleted indicates the size/number of data that have been completed and NTotal indicates the total size/number of data.  
+When the async operation completes, the Progress method returns an OperationProgress with ```Completed``` set as true. Then Velero will persist DUCR as well as any other items returned by DUP as ```itemToUpdate```.  
+Finally, then backup is as ```Completed```.  
+To help BIA Progress find the corresponding DUCR, the ```operationID``` is saved along with the DUCR as a label ```velero.io/async-operation-id```.  
+
+DUCRs are handled by the data movers, so how to handle them are totally decided by the data movers. Below covers the details of VBDM, plugging data movers should have their own actions and workflows.  
+
+**Persist DataUpload CR**  
+As mentioned above, the DUCR will be persisted when it is completed under the help of BIA V2 async operation finalizing mechanism.   
+This means the backup tarball will be uploaded twice, this is as the designed behavior of [general progress monitoring design][2].  
+
+Conclusively, as a result of the above executions:
+- A DataUpload CR is created and persisted to the backup tarball. The CR will be left there after the backup completes because the CR includes many information connecting to the backup that may be useful to end users or upper level modules.  
+- A snapshot as well as the objects representing it are created. For CSI snapshot, a VolumeSnapshot object and a VolumeSnapshotContent object is created. The DMP leaves the snapshot as well as its related objects to DM for further processing.  
+
+VBDM creates a Data Uploader Controller to handle the DUCRs in node-agent daemonset, therefore, on each node, there will be an instance of this controller. The controller connects to the backup repository and calls the uploader. Below are the VBDM actions.  
+
+**Acquire Object Lock**  
+**Release Object Lock**  
+There are multiple instances of Data Uploader Controllers and when a DUCR is created, there should be only one of the instances handle the CR.  
```

Comment:
The Data Uploader Controllers should be implemented in the Node-Agent, so the CR should be handled by the Node-Agent pod that shares the same node as the uploading volume, then there is only one candidate, and there is no need to have the lock.

Reply:
Actually, the Data Uploader Controller starts to process the DUCR once it is created (its phase is New or ""), then the controller creates the backupPVC, so it means the early phases before the backupPV is provisioned need the lock to make a consensus of which controller is responsible to create the backupPVC.